File: BirdTalkSdk/src/main/java/com/bird2fish/birdtalksdk/model/Topic.java
package com.bird2fish.birdtalksdk.model;

import com.bird2fish.birdtalksdk.db.TopicFlag;

public class Topic {
    private long tid;
    private long syncId;
    private long readId;
    private int type;
    private int data;
    private String title;
    private String icon;

    private long unReadCount;
    private MessageContent lastMsg;
    private boolean mute;  // 小写首字母，符合Java规范
    private boolean showHide; // 是否显示
    private boolean pinned;     // 置顶

    // 默认构造函数
    public Topic() {
        this(0L, 0L, 0L, 0, 1, "", "sys:11");
    }

    // 主构造函数
    public Topic(long tid, long syncId, long readId, int type, int data, String title, String icon) {
        this.tid = tid;
        this.syncId = syncId;
        this.readId = readId;
        this.type = type;
        this.data = data;
        this.title = title != null ? title : "";
        this.icon = icon != null ? icon : "sys:11";
        this.unReadCount = 0;
        this.lastMsg = null;
        this.mute = false;  // 默认不静音
        this.showHide = true;
        this.pinned = false;
    }

    // getter/setter
    public long getTid() {
        return tid;
    }

    public void setTid(long tid) {
        this.tid = tid;
    }

    public long getSyncId() {
        return syncId;
    }

    public void setSyncId(long syncId) {
        this.syncId = syncId;
    }

    public long getReadId() {
        return readId;
    }

    public void setReadId(long readId) {
        this.readId = readId;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public int getData() {
        return data;
    }

    public void setData(int mask) {
        this.data = mask;
        this.mute = (mask & TopicFlag.MUTE) != 0;
        this.showHide = (mask & TopicFlag.VISIBLE) != 0;
        this.pinned = (mask & TopicFlag.PINNED) != 0;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title != null ? title : "";
    }

    public String getIcon() {
        return icon;
    }

    public void setIcon(String icon) {
        this.icon = icon != null ? icon : "sys:11";
    }

    public long getUnReadCount() {
        return unReadCount;
    }

    public void setUnReadCount(long unReadCount) {
        this.unReadCount = unReadCount;
    }

    public MessageContent getLastMsg() {
        return lastMsg;
    }

    public void setLastMsg(MessageContent lastMsg) {
        this.lastMsg = lastMsg;
    }

    public boolean getMute() {
        return this.mute;
    }

    public boolean getPinned(){
        return this.pinned;
    }

    public boolean getShowHide(){
        return this.showHide;
    }
    public void setMute(boolean mute) {
        this.mute = mute;
        updateFlag(TopicFlag.MUTE, mute);
    }

    public void setPinned(boolean pinned) {
        this.pinned = pinned;
        updateFlag(TopicFlag.PINNED, pinned);
    }

    public void setShowHide(boolean visible) {
        this.showHide = visible;
        updateFlag(TopicFlag.VISIBLE, visible);
    }

    private void updateFlag(int flag, boolean enable) {
        if (enable) {
            data |= flag;
        } else {
            data &= ~flag;
        }
    }

    public long getTm(){
        if (this.lastMsg == null){
            return 0L;
        }

        return this.lastMsg.getTm();
    }

    @Override
    public String toString() {
        return "Topic{" +
                "tid=" + tid +
                ", syncId=" + syncId +
                ", readId=" + readId +
                ", type=" + type +
                ", visible=" + data +
                ", title='" + title + '\'' +
                ", icon='" + icon + '\'' +
                ", unReadCount=" + unReadCount +
                ", lastMsg=" + (lastMsg != null ? lastMsg.toString() : "null") +
                ", mute=" + mute +
                '}';
    }
}

File: BirdTalkSdk/src/main/java/com/bird2fish/birdtalksdk/model/Topic.kt
package com.bird2fish.birdtalksdk.model
import com.bird2fish.birdtalksdk.pbmodel.MsgOuterClass
import com.bird2fish.birdtalksdk.pbmodel.MsgOuterClass.ChatType

/**
 * 会话 / 话题实体
 * 数据库中有2个表,p_topic和g_topic
 * 分别存储;
 */
open class Topic(

        /* ================= 表字段 ================= */

        var tid: Long = 0L,
        var syncId: Long = 0L,
        var readId: Long = 0L,

        var type: Int = 0,
        private var data: Int = 1,            // flag 位（VISIBLE 默认 1）

        title: String = "",
        icon: String = DEFAULT_ICON
) {


    @Volatile
    private var _chatSession: ChatSession? = null

    fun asChatSession(): ChatSession {
        return _chatSession ?: synchronized(this) {
            _chatSession ?: ChatSession(this).also {
                _chatSession = it
            }
        }
    }

    /* ================= 展示字段 ================= */

    open var title: String = title.ifEmpty { "" }
    set(value) {
        field = value.ifEmpty { "" }
    }

    open var icon: String = icon.ifEmpty { DEFAULT_ICON }
    set(value) {
        field = value.ifEmpty { DEFAULT_ICON }
    }

    /* ================= 运行态字段（非表） ================= */

    var unReadCount: Long = 0
    var lastMsg: MessageContent? = null

    var mute: Boolean = false
        private set

    var showHide: Boolean = true
    private set

    var pinned: Boolean = false
    private set

    init {
        // 根据 data 同步状态位
        syncFlagsFromData()
    }

    /* ================= flag 同步 ================= */

    fun setData(mask: Int) {
        data = mask
        syncFlagsFromData()
    }

    fun getData(): Int = data

    private fun syncFlagsFromData() {
        mute = (data and TopicFlag.MUTE) != 0
        showHide = (data and TopicFlag.VISIBLE) != 0
        pinned = (data and TopicFlag.PINNED) != 0
    }

    private fun updateFlag(flag: Int, enable: Boolean) {
        data = if (enable) {
            data or flag
        } else {
            data and flag.inv()
        }
        syncFlagsFromData()
    }

    /* ================= 对外控制方法 ================= */

    fun setMute(enable: Boolean) {
        updateFlag(TopicFlag.MUTE, enable)
    }

    fun setPinned(enable: Boolean) {
        updateFlag(TopicFlag.PINNED, enable)
    }

    fun setShowHide(visible: Boolean) {
        updateFlag(TopicFlag.VISIBLE, visible)
    }

    /* ================= 便捷方法 ================= */

    fun getTm(): Long {
        return lastMsg?.tm ?: 0L
    }

    override fun toString(): String {
        return "Topic(" +
                "tid=$tid, " +
                "syncId=$syncId, " +
                "readId=$readId, " +
                "type=$type, " +
                "data=$data, " +
                "title='$title', " +
                "icon='$icon', " +
                "unReadCount=$unReadCount, " +
                "lastMsg=${lastMsg ?: "null"}, " +
                "mute=$mute, " +
                "pinned=$pinned, " +
                "showHide=$showHide" +
                ")"
    }

    // 默认的
    companion object {
        const val DEFAULT_ICON = ""
        val CHAT_P2P = ChatType.ChatTypeP2P.number
        val CHAT_GROUP = ChatType.ChatTypeGroup.number
        var CHAT_SYSTEM = 0
    }


}
